### DISCUSSION TITLE
Creating a windows binary and running from inside WSL cannot be accessed on localhost endpoint

### DISCUSSION THREAD

**Discussion by PylotLight** [pocketbase/pocketbase#4367] [Q&A]
Let me know if this is out of scope, but wasn't sure where else to ask about this.
1. Create a binary with the below code and build for linux: works
2. Create a binary with same code built as windows and run from WSL: does not work and is not accessible.
Note: 
- This may be due to the nuances of WSL networking, however I tested 4 other http web servers run in the same manner which all worked fine when running from windows, so its only pocketbase that I've observed this behavior for.
- This has also been observed using the prebuilt pocketbase exes, same result for linux working but windows does not.
- See also image for simple http server used which confirms its not a WSL only issue as its also running from linux and it works fine.
- Also any normal http server built in go also works fine, see image where top pocketbase code is not accessible but the bottom is.
Code:

```go
package main

import (
    "fmt"

    "github.com/pocketbase/pocketbase"
    "github.com/pocketbase/pocketbase/apis"
)

func main() {
    pocketBaseApp := pocketbase.New()
    pocketBaseApp.Bootstrap()

    serv, err := apis.Serve(pocketBaseApp, apis.ServeConfig{
        HttpAddr:        "127.0.0.1:8090",
        ShowStartBanner: false,
    })
    fmt.Println("test", err.Error(), serv.Addr)
}
``` 
![image](https://github.com/pocketbase/pocketbase/assets/7006124/623b81e0-0b01-420a-8089-b5b0651aa207)
![image](https://github.com/pocketbase/pocketbase/assets/7006124/93a09152-af7f-4b65-a0c5-2a3b9bedeea4)
![image](https://github.com/pocketbase/pocketbase/assets/7006124/4c2d8ffa-0a98-435b-a7b6-efe71666717f)
![image](https://github.com/pocketbase/pocketbase/assets/7006124/6c8f8d41-0aa3-44e0-911e-2affaca5e84a)

**Comment by ganigeorgiev** [comment]
_Please don't post code as screenshots._

I'm not familiar with Windows or WSL but I don't think your issue is PocketBase related.

Note that binding to `:3333` is not the same as binding to `127.0.0.1:8090`. When no address is specified it binds to all available anycast and unicast IP addresses on the local system.

**Comment by PylotLight** [comment]
> _Please don't post code as screenshots._
> 
> I'm not familiar with Windows or WSL but I don't think your issue is PocketBase related.
> 
> Note that binding to `:3333` is not the same as binding to `127.0.0.1:8090`. When no address is specified it binds to all available anycast and unicast IP addresses on the local system.
> 
> 
> 
> 

Oh I've tried all of that already. Also note the pocketbase binary attempt which uses the default binding. 
Also I did post code and screenshots, so it's all included.

**Reply by ganigeorgiev** [reply]
I've just tested the prebuilt windows amd64 executable on Windows 10 VM (running directly from 
the default cmd) and it is working fine for me so something else is going on in your environment but I don't think I can help you.

Since I'm not sure that I understand how you are testing, nor how WSL works, if you are trying to cross-compile windows->linux (or the opposite) you'll have to specify the `GOOS` and `GOARCH` env variables. The pure GO SQLite driver currently supports the following platforms:

```
darwin  amd64
darwin  arm64
freebsd amd64
freebsd arm64
linux   386
linux   amd64
linux   arm
linux   arm64
linux   ppc64le
linux   riscv64
linux   s390x
windows amd64
windows arm64
```

**Reply by ganigeorgiev** [reply]
_As an alternative, if you haven't already tried this, you can replace `127.0.0.1:8090` with `0.0.0.0:8090` to bind to all available IPv4 interfaces, but again I don't know how WSL works. PocketBase uses the standard Go `net/http` so it is expected the PocketBase server to behave the same as in your plain Go example as long as the same binding address is used._

**Reply by PylotLight** [reply]
Ye defs tried all that on multiple machines. All W11 though but that shouldn't matter here. 
Hmm. 

Note sure why you're mentioning cross compiling, as building the binaries themselves is no issue here. 

I'll try test on more environments and distros perhaps.

**Reply by PylotLight** [reply]
Ah if I leave my windows build running it errors: The code is same as above.
```go
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x15c2667]

goroutine 1 [running]:
main.main()
        /home/light/test-proj/main.go:18 +0xe7
``` 
![image](https://github.com/pocketbase/pocketbase/assets/7006124/88e5cec5-23b3-42b1-9210-e4648de15953)

**Reply by ganigeorgiev** [reply]
Since it fails on line 18 make sure that the variables you print (`err` and `serv`) are not nil.

I'm lacking the full picture of your use case, why do you initialize manually `apis.Serve`, what is the purpose of the "Simple web server" screenshot (~is this from a wails app?~ _it seems to be an Electron app but I'm even more confused how it is related to PocketBase_), why do you want to run a Windows binary from a WSL (note that you can run the exe directly from the Windows cmd and you don't need extra Linux virtualization), but I don't think your issue is PocketBase related as I'm not able to reproduce it locally on the free trial Windows 10 VM.

And again, we don't do anything more different than you plain Go example. The only difference would be the binding address you use.
If you think that it is a PocketBase issue feel free to create a PR and I'll try to review it but until proven otherwise, for me this is something related to your environment setup or you are not sharing the exact steps you are executing.

**Reply by PylotLight** [reply]
1. I've removed the print and now it just exits after like a minute but doesn't error as before at least.
![image](https://github.com/pocketbase/pocketbase/assets/7006124/470837e7-a798-478e-9164-ae649b29378c)
![image](https://github.com/pocketbase/pocketbase/assets/7006124/17daf702-9d42-4539-b999-c30487d1d0a3)


2. This was the code provided in other samples from other users, I've tried other variations of this like:
```go
...
app := pocketbase.New()
app.Bootstrap()
serveCmd := cmd.NewServeCommand(app, true)
serveCmd.Execute()
...
``` 
3. Simple web server demonstrates that network requests are available by apps run inside WSL so its not a WSL networking issue but app related because other web servers run fun, the technology that runs the server is irrelevant.
4. Why? That's a separate question and not really relevant to the issue, if we find that my use case is broken we can review that question but at this time I have my reasons.
5. I did look through the PB code and you're mostly correct that its basically a http web server which is why I'm so confused on the different behavior from a normal go http server spun up which works just fine as illustrated in the OP.
6. Also the latest screenshot literally entails the entire process + the code in the OP built as windows and linux binaries and run on the cli. (Also used the PB prebuilt unedited binaries on the main page which also have the same issue)
7. This is not at all resolved and feels like you still haven't understood the scenario?

**Reply by PylotLight** [reply]
Next code test:
```go
package main

import (
    "github.com/pocketbase/pocketbase"
    "github.com/pocketbase/pocketbase/apis"
)

func main() {
    pocketBaseApp := pocketbase.New()
    pocketBaseApp.Bootstrap()

    _, err := apis.Serve(pocketBaseApp, apis.ServeConfig{
        HttpAddr:        "127.0.0.1:8090",
        ShowStartBanner: false,
    })
    println("test", err.Error())
}
``` 

light@pc:~/test-proj$ ./pb.exe
test database is locked (5) (SQLITE_BUSY)

**Reply by ganigeorgiev** [reply]
I'll reopen the discussion in case someone else can help you.

I have no idea why the above code will trigger the `SQLITE_BUSY` error. It is possible to conflict with other applications interacting with the SQLite db file (Litesteam or another PocketBase process).

But once again it doesn't make sense why it will work with the plain Go example and not with PocketBase unless you are using a different binding address.

**Comment by PylotLight** [comment]
> I have no idea why the above code will trigger the `SQLITE_BUSY` error. It is possible to conflict with other applications interacting with the SQLite db file (Litesteam or another PocketBase process).
> 
> But once again it doesn't make sense why it will work with the plain Go example and not with PocketBase unless you are using a different binding address.

Well obviously the main difference is the presence of a database setup and migrations in the background which a basic http server does not have. So my next test would be to remove the migrations step from the .Serve() part as it seems to me that there's some issue with the DB part that it's trying to setup but can't,.

**Reply by ganigeorgiev** [reply]
By comparing with the plain Go example I meant that initially you didn't report any error and I assume that the `SQLITE_BUSY` is something else. But in any case I'll stop engage to avoid further confusion.

**Comment by PylotLight** [comment]
Sorry about sucking with threads above I keep getting distracted and not doing proper replies.
Anyway I've been adding some logging and found issues with connection to the db context via some print logging which I'm still figuring out the cause for, because migrations is failing the web server does not start which is why I've been having trouble, now I just need to figure out this db issue, if there's anyone that wants to advise on that part let me know :P
See current status:
![image](https://github.com/pocketbase/pocketbase/assets/7006124/e4f92445-c767-4219-81ac-5e73253ab5b8)
![image](https://github.com/pocketbase/pocketbase/assets/7006124/b37f1036-0d19-47a2-be5e-6e74e28608ff)

**Reply by PylotLight** [reply]
My current guess is that running a windows binary makes it use windows paths for the db which doesn't work when run inside wsl, so we may need to add path translation in this case. I'll see if I can figure out how do make that happen (notice the backslash instead of forwardslash)
![image](https://github.com/pocketbase/pocketbase/assets/7006124/ed136ef5-7ea4-4d09-978f-72bd38c33a82)
![image](https://github.com/pocketbase/pocketbase/assets/7006124/258c2368-b8ed-473c-9f91-cdfd3bdbb6c1)

Hmm that didn't seem to work, removing the pb_data subfolder putting the .db files in root of proj dir still has same issue. Bit lost on that one now, but I still think there's something wrong with the setting up/connection of these db files, so the database is locked is a bit of a lie and it actually just can't create them properly or something.

Can't debug in windows so just have the print statements, but observations:
1. ~~No pb_sqlite3 driver is registered, only sqlite3~~ Ah windows runs nocgo that's why.
2. Linux uses full path while windows only uses direct relative folder.
3. Ah windows runs nocgo that's why.

Linux:
![image](https://github.com/pocketbase/pocketbase/assets/7006124/de09fe59-07c8-44d8-af49-b50b391f77bf)
![image](https://github.com/pocketbase/pocketbase/assets/7006124/08623559-1bc3-4524-a68a-d95ef836aa4f)

Windows:
![image](https://github.com/pocketbase/pocketbase/assets/7006124/e0aaf228-200c-449b-a66f-92f4988fa999)


Confirmed that windows can't seem to connect to the file handles properly when run through WSL:
![image](https://github.com/pocketbase/pocketbase/assets/7006124/91a53146-7ead-43a3-8da8-c44160cf7305)

Image is showing a file handle monitor `fuser` being run where previously in linux mode it shows PID, but for windows no file handle is established. If I can figure out why it can't connect to the db file then I'm hoping that will help.

**Reply by ganigeorgiev** [reply]
@PylotLight I cannot help you and I think you are wasting your time (or at least I don't understand what you are trying to do).

While Go can cross-compile, the produced binaries are OS/architecture dependent (_this may change if support for something like [`GOOS=cosmo`](https://github.com/golang/go/issues/51900) is added in the future_).

If you want to run the binary from WSL try with the Linux version.
The Windows binary is expected to work correctly only when executed natively (aka. without WSL). And even if you resolve your issue with the `SQLITE_BUSY` error, which I'm not sure that it is caused by the WSL interop, there is no guarantee that everything else will operate properly because there are a lot of places in the Go std library and the PocketBase dependencies where platform specific code is loaded during build (_or through various runtime conditional checks_).

If you still think that it is a PocketBase issue, feel free to open a PR with the suggested fix and I'll try to review it in a timely manner, but until then I'll close discussion to avoid further spamming the watchers with notifications.

**Reply by PylotLight** [reply]
The watchers is just you, so you could unsubscribe, but the purpose of this thread was to document my findings. 
So far it does seem like a WSL related issue in terms of file interop as the SQLite browser run from windows also can't open those DB files. 

I purposefully don't want the linux version as I'm building a windows app not a linux app, so running linux really doesn't help me here. 

But yes at this point it's not a pocketbase specific issue, more on the WSL/sql side for file access as far as I can tell. 
https://github.com/microsoft/WSL/issues/4689

One quick question in case you check this thread again, how can we specify datadir so we can choose a windows path instead of linux path where the binary is located?

**Reply by hesusruiz** [reply]
As @ganigeorgiev has said, I do not understand what exactly you are trying to achieve. WSL is intented to allow runing linux executables almost natively, while windows binaries are intended to be run in the Windows environment. You are testing the interoperability of WSL and Windows file systems (among other things), which is OK for symple things but breaks for complex requirements.

There are many Windows binaries out there which do not work in WSL, especially database systems, because the special and tricky file system features they expect from the underlying operating system. Try running the Windows version of MySQL in WSL.

The recommendation from MS (I do not remember where I read it) is essentially: Linux binaries running in WSL should access the WSL file system, if possible, and the Windows binaries running in the Windows environment should access the Windows file system. Both are "visible" from the other environment, but crossing boundaries should be done only in special (and simpe) situations, like copying friles from one environment to the other (and even in this case, there may be problems with permissions for example).

If you build Pocketbase in WSL and run it in WSL, you should not get these problems. The same if you build Pocketbase in Windows and run it in Windows.

Hope it helps.
